### PR TITLE
Alexander Lind

### PR DESCRIPTION
**Pull request by AlexLind**

### CODE REVIEW

File: src/controllers/movies.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  try {
+    const createdMovie = await prisma.movie.create({
+      data: {
+        title: title,
+        runtimeMins: runtimeMins,
+        screenings: screenings ? includeScreenings : undefined,
+      },
+      include: { screenings: true },
+    });
+
```
Comment (vherus) [inline src/controllers/movies.js:53]: We could avoid the use of `let` here with something like this:

```js
const data = { title, runtimeMins }

if (screenings) {
    data.screenings = screenings
}

try {
    const createdMovie = await prisma.movie.create({
        data,
        include: { screenings: true }
    })
```

File: src/controllers/movies.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  const queryParam = req.params.id.replace("%", " ");
+  let id
+  let title
+  if (isNaN(Number(queryParam))) {
+    id = 9999;
+    title = queryParam
+  } else {
+    id = Number(queryParam);
+    title = 'Not a real movie, help me with this functionality!!!'
+  }
```
Comment (vherus) [inline src/controllers/movies.js:79]: What's going on here 🤔

File: test/api/extensions/customer-ext.spec.js [hunk 3]
```diff
@@ -1,56 +1,154 @@
...
-        })
-    })
-})
+const supertest = require("supertest");
+const app = require("../../../src/server.js");
+const { createCustomer } = require("../../helpers/createCustomer.js");
+const { createScreen } = require("../../helpers/createScreen.js");
+const { createMovie } = require("../../helpers/createMovie.js");
+
+// describe("Customer Endpoint", () => {
```
Comment (vherus) [inline test/api/extensions/customer-ext.spec.js:7]: Did this test stop passing after working on the exercise?

Comment (AlexLind) [reply test/api/extensions/customer-ext.spec.js]: I never got to work on the customer endpoint in the extensions. I have removed this test file in the final commit.

File: src/controllers/movies.js (hunk 2 above)
Comment (AlexLind) [reply src/controllers/movies.js]: A bad piece of code to handle being able to search for a movie either by ID (int) or title (String). It's passed as a string in the params so need to typecast it if it's a number, or leave it alone if it's a string.

File: src/controllers/movies.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+  const createdMovie = await prisma.movie.create({
+    data: {
+      title: title,
+      runtimeMins: runtimeMins,
+      screenings: screenings ? includeScreenings : undefined,
+    },
+    include: { screenings: true },
+  });
```
Comment (vherus) [inline src/controllers/movies.js:49]: This is good, a small improvement here would be to remove the `let includeScreenings` and attach the createMany to the data object in the `if` statement, like this:

```js
  if (!title || !runtimeMins) {
    throw new MissingFieldsError();
  }

 const data = { titletitle, runtimeMins }

  if (screenings) {
    data.screenings = {
      createMany: {
        data: screenings.map(({ screenId, startsAt }) => ({
          screenId,
          startsAt: new Date(startsAt),
        })),
      },
    };
  }

  const createdMovie = await prisma.movie.create({
    data,
    include: { screenings: true },
  });
```

File: src/server.js [hunk 5]
```diff
@@ -13,9 +17,34 @@ app.use(express.json());
...
+// Error handling
+app.use((e, req, res, next) => {
+    if (e instanceof Prisma.PrismaClientKnownRequestError) {
+        if (e.code === "P2002") {
+            return res.status(409).json({ error: "A customer with the provided email already exists" })
+        }
+        if (e.code === "P2000") {
+            return res.status(409).json({ error: "Value provided was too long" })
+        }
+    }
```
Comment (vherus) [inline src/server.js:38]: Not necessary, but this kind of situation is one of those rarer cases where a switch statement might come in handy. I don't like to use them very often, but sometimes they can be useful:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch

```js
if (e instanceof Prisma.PrismaClientKnownRequestError) {
  switch (e.code) {
    case "P2002":
      return res.status(409).json({ error: "A customer with the provided email already exists" })
      break
    
    case "P2013"
      ...
  }
}
```

**Comment by vherus** [comment]
Great work on this!

**Pull request closed without merging**